packages/sdk/src/protocols/base_protocol.ts

```diff
@@ -1,15 +1,206 @@
+import type { Peer, PeerId } from "@libp2p/interface";
+import { ConnectionManager } from "@waku/core";
+import { BaseProtocol } from "@waku/core/lib/base_protocol";
 import { IBaseProtocolSDK } from "@waku/interfaces";
+import { Logger } from "@waku/utils";
 
 interface Options {
   numPeersToUse?: number;
+  maintainPeersInterval?: number;
 }
 
 const DEFAULT_NUM_PEERS_TO_USE = 3;
+const DEFAULT_MAINTAIN_PEERS_INTERVAL = 30_000;
 
 export class BaseProtocolSDK implements IBaseProtocolSDK {
-  public readonly numPeers: number;
+  public readonly numPeersToUse: number;
+  private peers: Peer[] = [];
+  private maintainPeersIntervalId: NodeJS.Timeout | undefined;
+  log = new Logger("sdk:base-protocol");
 
-  constructor(options: Options) {
-    this.numPeers = options?.numPeersToUse ?? DEFAULT_NUM_PEERS_TO_USE;
+  maintainPeersLock = false;
+
+  constructor(
+    protected core: BaseProtocol,
+    private connectionManager: ConnectionManager,
+    options: Options
+  ) {
+    this.log = new Logger(`sdk:${core.multicodec}`);
+    this.numPeersToUse = options?.numPeersToUse ?? DEFAULT_NUM_PEERS_TO_USE;
+    const maintainPeersInterval =
+      options?.maintainPeersInterval ?? DEFAULT_MAINTAIN_PEERS_INTERVAL;
+
+    this.startMaintainPeersInterval(maintainPeersInterval).catch((error) => {
+      this.log.error("Error starting maintain peers interval:", error);
```

Comment:
I don't think we should re-throw here as: error won't be handled up the tree, it is not actionable up there and in general (in our codebase too) we assume `ctor` do not throw.